### DISCUSSION TITLE
how to create reset link

### DISCUSSION THREAD

**Discussion by puraminy** [pallets-eco/flask-security#637] [General]
Hello,

Currently, I don't want the whole functionality of this package which is a bit complex for me. I just want to send to the user a link by which he can reset his password, I want the link and a form to reset. Do you have a minimal example for it?

**Comment by jwag956** [comment]
Almost all functionality from Flask-Security is optional - if you only set SECURITY_RECOVERABLE - you will have just a /login endpoint and a /reset endpoint.

If you don't want to use the FS package at all - look at the itsdangerous package - that is what FS uses to create the link - and as the data in the link put in some user identifier...

**Reply by puraminy** [reply]
Thank you very much, I already tried `itsdangerous` package to create the link. Since my website is in Persian and I have my own messages in Persian for email and etc, I implemented the email part too, unless your package support this part too. 

I mean if I use that setting should I implement `reset` endpoint and if yes, can you provide an small example of it?

**Reply by jwag956** [reply]
Flask-Security supports localizing all messages and templates - there are already many languages included in the package.

**Reply by puraminy** [reply]
Thanks, yes, now I am reading those documents and realized it. I'm implementing your basic examples ....

**Reply by jwag956** [reply]
You probably missed the line: fsqla.FsModels.set_db_info(db) in your main app.

**Comment by puraminy** [comment]
I got it running but after introducing `User` and `Role` models. However I already had a `Users` model (please note it is plural and also had other fields), if I drive it from `fsqla.FsUserMixin` it gives some errors like lacking of a relationship with roles, etc. I would like to know how can I use my current defined model?

**Reply by jwag956** [reply]
If you already have a User model defined - then the easiest thing to do is not use fsqla and manually add the required fields to your existing model

**Reply by puraminy** [reply]
I tried to follow this example at end of this doc, and defined them as you did (plus my own fields and names):

https://flask-security-too.readthedocs.io/en/stable/two_factor_configurations.html#adding-sms

And commented these lines:

```
#from flask_security.models import fsqla_v2 as fsqla
#fsqla.FsModels.set_db_info(db)
from apps.authentication.models import Users, Role

user_datastore = SQLAlchemyUserDatastore(db, Users, Role)
```

is it what you mean? Now, what? I also have had register and login, logout endpoints before, should I change them? 

I am going now to see how can I use your features...

**Reply by puraminy** [reply]
The first error is that it doesn't register the user with the error:

```
sqlalchemy.exc.OperationalError: (MySQLdb.OperationalError) (1048, "Column 'username' cannot be null")
[SQL: INSERT INTO `Users` (email, username, active, fs_uniquifier, confirmed_at, tf_phone_number, tf_primary_method, tf_totp_secret, password, failed_logins, logins, mobile, image, cover) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)]
[parameters: (None, None, None, None, None, None, None, None, None, 0, 0, None, '/static/assets/img/anon-user.jpg', '')]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```

It was working before with this code, and `reqform` contains required parameters:
```
@blueprint.route('/register', methods=['GET', 'POST'])
def register():
    reqform = request.form.copy()
    reqform = convert(reqform, del_crf=False)
    create_account_form = CreateAccountForm()
    if 'register' in reqform:

        username = reqform['username']
        email = reqform['email']
        mobile = reqform['mobile']
        password = reqform['password']
        # .... doing some validation and 
        user = Users(**reqform, logins=1)
        db.session.add(user)
        db.session.commit()
```
The error raises from `db.session.commit()` line as if it ignores my data provided in previous lines...

**Reply by jwag956** [reply]
If your user model requires username then your form has to validate for that. If username is optional - then make sure your SQL column is set to 'nullable'

**Reply by puraminy** [reply]
The username was provided in the code above, if I put an assertion before the last line (commit), 

```
        assert False, reqform
        user = Users(**reqform, logins=1)
        db.session.add(user)
        db.session.commit()

```
it gives:

`AssertionError: {'csrf_token': 'IjQ0NDMyNjk3MDM3YTNhODFhYzA2YTZiZGI4MjgwOTUwODcyMTlkZjMi.YtWHtg.rThpLYJr3PudYRItHc4FTyU1KsQ', 'username': '3149091578', 'mobile': '9133923475', 'email': 'alirezayee@chmail.ir', 'password': '0123456789', 'register': ''}
`

Did you noted my `register` code? Should I change it? It was working, I am now confused...

**Reply by jwag956** [reply]
So if you look at the SQL error - NONE of your parameters have been set - I would print out the 'user' variable. If you show your models file (especially the Users model) I can take a look.
Hopefully you have an IDE such as PyCharm which would let you quickly pop in the debugger and trace out what is happening.

**Reply by puraminy** [reply]
These are my models that I have adopted based on comments above. yes if I assert `user` it's `None`.
I use `vim` editor and doesn't have IDE 
```
from flask_security import Security, SQLAlchemyUserDatastore, \
    UserMixin, RoleMixin, auth_required 
    
roles_users = db.Table('roles_users',
    db.Column('user_id', db.Integer(), db.ForeignKey('Users.id')),
    db.Column('role_id', db.Integer(), db.ForeignKey('role.id')))
  
class Role(db.Model, RoleMixin):
  id = db.Column(db.Integer(), primary_key=True)
  name = db.Column(db.String(80), unique=True)
  description = db.Column(db.String(255))
  
class Users(db.Model, UserMixin):
    __tablename__ = 'Users'
    __table_args__ = {'mysql_engine':'InnoDB','mysql_charset':'utf8','mysql_collate':'utf8_general_ci'} 
    id = db.Column(db.Integer, primary_key=True)
    email = col(db.String(64),"ﭗﺴﺗ ﺎﻠﮑﺗﺭﻮﻨﯿﮑﯾ", unique=False)
    email.info["ltr"] = True

    username = col(db.String(64),"ﺶﻣﺍﺮﻫ ﻢﻠﯾ", unique=True, nullable=False)
    password = db.Column(db.String(255))
    active = db.Column(db.Boolean())
    fs_uniquifier = db.Column(db.String(255), unique=True, nullable=False)
    confirmed_at = db.Column(db.DateTime())
    roles = db.relationship('Role', secondary=roles_users,
                            backref=db.backref('users', lazy='dynamic'))
    tf_phone_number = db.Column(db.String(128), nullable=True)
    tf_primary_method = db.Column(db.String(64), nullable=True)
    tf_totp_secret = db.Column(db.String(255), nullable=True)

    failed_logins = col(db.Integer, default=0)
    logins = col(db.Integer, default=0)
    mobile = col(db.String(64),"ﺖﻠﻔﻧ ﻪﻣﺭﺎﻫ", unique=False, nullable=False)
    image = col(db.String(50),
            default="/static/assets/img/anon-user.jpg",
            unique=False)
    cover = col(db.String(50), default="", unique=False)
    def __init__(self, **kwargs):
        for property, value in kwargs.items():
            # depending on whether value is an iterable or not, we must
            # unpack it's value (when **kwargs is request.form, some values
            # will be a 1-element list)
            if hasattr(value, '__iter__') and not isinstance(value, str):
                # the ,= unpack of a singleton fails PEP8 (travis flake8 test)
                value = value[0]

            #if property == 'password':
            #    value = hash_pass(value)  # we need bytes here (not plain str)
    def __repr__(self):
        return str(self.username)


                                                                          103

**Reply by jwag956** [reply]
your __init__ isn't initializing properties that AREN'T iterables.

I strongly suggest you download and use the community version of PyCharm.

**Reply by puraminy** [reply]
Thanks, now I reverted to my own project and I guess my code with `itsdangerious` to create link is enough for now. I just remember when I removed the `init` part, the user was still `None`, so the error probably exists there and is more complex to continue...
Thanks anyways, If I switched back to use your package I will discuss the error later.
